### ISSUE TITLE
[Help] Detect when "Step" is finished parsing entire file?

### ISSUE THREAD

**Issue by ewliang** [mholt/PapaParse#718]
So I have a csv file which can have varying row counts and potentially columns.
However, I am only using papaparse to parse certain named headers.
All of this is done via the "Step" function in papaparse.
From there, I'm using read/write streams to write the parsed data into a results file with a new column to state whether it was successfully parsed and written to database via mongoose.
Now my issue is is there a way to detect all the N variable row counts have been parsed? 
The results file will differ in size and column count, but row count will remain exactly the same as the original file.
Basically, once the results file is generated, I need some way for the code to know that step has parsed the last row in the uploaded csv file. Then once it's done, it will tell nodejs that ok, it's time to send the results.csv file to the client frontend for the user to download.

Again, I'm using the "Step" function in papaparse.

Some sample code I have are along the lines of:
```
var reader = createReadStream(filepath); //UTF8 Encoding`
var writer = createWriteStream(....results.csv); //Create a empty results.csv to write to.`
for loop // use "writer" to write array of custom headers to results.csv prior to papaparse occurs.
Papa.parse(reader, {
step: function(row) {
    if(the columns have certain data that passes certain requirements) {
        var myObject = new Goods();
                   //populate myObject with the data in the current row
                   myObject.save(err => { // save to mongoose mongodb database
                           writer.write(myObject) // write to results file with status column is true.
                           if error, writer.write(status column is false)
                   })
            } else {
                   var myObject = new Goods();
                   //populate myObject with the data in the current row
                   writer.write(myObject) //write to results file with status column is false
            }
      }
});
```
//Now i need a way so my program knows step function has processed the last row so that it knows writer is also done from the mongoose callback so i can send the results file after everything is finished. I tried on('end') for both reader and writer, but they fire before the results file gets data written in it. Hence why i need a way to know when step function reaches the end. Thanks!

Thank you!

**Comment by jseter** [comment]
there are two additional hooks on the config option.  The complete and error callbacks can be specified to trigger the desired action when parsing completes.

**Comment by ewliang** [comment]
> there are two additional hooks on the config option. The complete and error callbacks can be specified to trigger the desired action when parsing completes.

Hi jseter,

Thank you for replying, but I am already aware of the complete/error callbacks that you mentioned, but in my case, because I am using the step function. Based on my understanding of the documentation, I believe the complete function is not the answer to my problem. I've also tried using the complete with step before at an attempt to address my issue, but sadly, it doesn't provide the result that I want (unless I am misunderstanding something here...)

**Comment by jseter** [comment]
The complete or error callback is fired after the file is done parsing, which means the last row has definitely been parsed and delivered to step.
As I understand your issue; it is more of a standard asynchronous programming issue than it is about the parser.
You have a chain of 3 separate asynchronous elements which you need to properly coordinate to solve the problem. The first issue I see is that you are firing Goods.save repeatedly with no back-pressure handling.  The step callback will keep firing until it finishes the file, and will start the next save() before the previous call to save() has completed.  This will keep queuing up asynchronous work until you run out of resources (terminating node) or you finish parsing the file.  As you see the step function does not have a callback, so it will not wait for these asynchronous items to complete before it continues to parsing the next line.
Additionally there is no back-pressure handling on the write() method call.  The same problem occurs that you will keep writing until you are out of resources, so you will need to coordinate this resource as well.
When it finishes parsing the file your app will eventually finish all of these queued asynchronous tasks but at some point after the parser has fired the complete callback.  Since you didn't track the state of your writing, you have no idea when everything is written.

If the files are small you could just let the parser return the full results in memory and deal with the asynchronous saving/writing separately.  However if the files are large or streaming is the right use case; What you could possibly do is modify the step code to wait for the asynchronous code to complete before it continues parsing the file.  That way when the parser's complete or error callback fires to indicate it finished parsing the file you know you are done because there is no outstanding asynchronous work in the background.  At that time you'll have to close your write stream and let it flush to disk so that it is usable for your upload.

```
// no warranty on this concept code
step: (row, parser) => {
    parser.pause();
    stepAsync(row, parser, (err) => {
        if(err) {
              // handle the error.
              return;
        }
        // queue to continue the parsing without blocking here and growing the call stack.
        queueMicrotask(() => { parser.resume(); });
    }
},
complete: () => {
    writer.end();
    writer.once('finish', () => {
        // we are done, so safe to use the file now.
    });
    writer.once('error', (err) => {
        // handle the error.
    });
}

...
function queueMicrotask(fn) {
    // assuming node.
    process.nextTick(fn);
    // if we have promises
    // Promise.resolve().then(() => { fn(); });
    // or we can use setTimeout
    // setTimeout(() => { fn(); }, 0);
}

function stepAsync(row, parser, cb) {
    if(the columns have certain data that passes certain requirements) {
        var myObject = new Goods();
        //populate myObject with the data in the current row
        myObject.save((err) => { // save to mongoose mongodb database
            if(err) {
                cb(err);
                return;
            }
            writeAsync(myObject, cb);
        });
    } else {
        var myObject = new Goods();
        //populate myObject with the data in the current row
        writeAsync(myObject, cb);
    }
}

function writeAsync(obj, cb) {
    const needDrain = !writer.write(myObject);
    if(needDrain) {
         writer.once('drain', cb);
    } else {
        queueMicrotask(cb);
    }
}
```

**Comment by ewliang** [comment]
@jseter 

Thank you for the response.
I managed to figure out the issue. 
Turns out the code I had was parsing the rows in a non-synchronous manner which ended up being so fast that there was no way for my code to detect when the parsing was finished.
All I had to do was include the "parser" argument in my step() function and just call pause it after I got the data I needed, then parser.resume() in each of my nested conditions/callbacks specific to my case.... and the whole thing worked like a charm!

**Comment by johanwesteinde** [comment]
Hi @ewliang, I'm running into a similar issue. I'm looking to call another function and pass the parsed data only once all the data has been parsed (via `step` function) and I'm not sure how to tell when I've reached the end of the file. My rough code structure is this: 

```
step: (results, parser) => {
        if (results is not a duplicate of a previous row) {
          //add data to array, carry on
          data += results.data;
        }

        //once all rows have been streamed, I want to call a function and pass the parsed data to it. I don't want this called after each row has finished streaming, only at the end of file. 
        handleData(data);
      }
```
